src/plugins/push/pushactivation.ts

```diff
@@ -0,0 +1,605 @@
+        headers = machine.rest.Defaults.defaultPostHeaders(rest.options, { format }),
+        params = {};
+
+      if (rest.options.headers) machine.rest.Utils.mixin(headers, rest.options.headers);
+
+      if (rest.options.pushFullWait) machine.rest.Utils.mixin(params, { fullWait: 'true' });
+
+      const requestBody = machine.rest.Utils.encodeBody(body, rest._MsgPack, format);
+
+      machine.rest.Resource?.post(rest, '/push/deviceRegistrations', requestBody, headers, params, null, true)
+        .then((response) => {
+          const deviceDetails = response.unpacked
+            ? response.body
+            : rest.DeviceDetails?.fromResponseBody(response.body as any, rest._MsgPack, format);
+          machine.handleEvent(new GotDeviceRegistration(deviceDetails as DeviceRegistration));
+        })
+        .catch((err) => {
+          machine.handleEvent(new GettingDeviceRegistrationFailed(err as ErrorInfo));
+        });
+    }
+
+    return WaitingForDeviceRegistration;
+  } else if (event instanceof GettingPushDeviceDetailsFailed) {
+    machine.callActivatedCallback(event.reason);
+    return NotActivated;
+  }
+  return null;
+};
+
+const WaitingForDeviceRegistration = function (machine: ActivationStateMachine, event: ActivationEvent) {
+  if (event instanceof CalledActivate) {
+    return WaitingForDeviceRegistration;
+  } else if (event instanceof GotDeviceRegistration) {
+    const device = machine.rest.device as LocalDevice;
+    device.deviceIdentityToken = event.tokenDetails.token;
+    device.persist();
+    machine.callActivatedCallback(null);
+    return WaitingForNewPushDeviceDetails;
+  } else if (event instanceof GettingDeviceRegistrationFailed) {
+    machine.callActivatedCallback(event.reason);
+    return NotActivated;
+  }
+  return null;
+};
+
+const WaitingForNewPushDeviceDetails = function (machine: ActivationStateMachine, event: ActivationEvent) {
+  if (event instanceof CalledActivate) {
+    machine.callActivatedCallback(null);
+    return WaitingForNewPushDeviceDetails;
+  } else if (event instanceof CalledDeactivate) {
+    machine.deregister();
+    return WaitingForDeregistration(WaitingForNewPushDeviceDetails);
+  } else if (event instanceof GotPushDeviceDetails) {
+    machine.updateRegistration();
+    return WaitingForRegistrationSync;
+  }
+};
+
+const WaitingForRegistrationSync = function (machine: ActivationStateMachine, event: ActivationEvent) {
+  if (event instanceof CalledActivate) {
+    machine.callActivatedCallback(null);
+    return WaitingForRegistrationSync;
+  } else if (event instanceof RegistrationUpdated) {
+    return WaitingForNewPushDeviceDetails;
+  } else if (event instanceof SyncRegistrationFailed) {
+    machine.callUpdateRegistrationFailedCallback(event.reason);
+    return AfterRegistrationUpdateFailed;
+  }
+  return null;
+};
+
+const AfterRegistrationUpdateFailed = function (machine: ActivationStateMachine, event: ActivationEvent) {
```

Comment:
I think it's called `AfterRegistrationSyncFailed` in [spec](https://sdk.ably.com/builds/ably/specification/main/features/#RSH3f)?